### PR TITLE
feat(newsletter): added hubspot driver

### PR DESCRIPTION
**Pull request by Nolan-Arnold**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Fixes: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

## Other information

### CODE REVIEW

File: libs/newsletter/package.json [hunk 1]
```diff
@@ -13,6 +13,9 @@
     "type": "git",
     "url": "https://github.com/graycoreio/daffodil"
   },
+  "publishConfig": {
+    "directory": "../../dist/newsletter"
```
Comment (damienwebdev) [inline libs/newsletter/package.json]: Why is this needed?

File: libs/newsletter/src/driver/hubspot/transformers/newsletter.transformer.spec.ts [hunk 2]
```diff
@@ -1,5 +1,4 @@
 import { DaffNewsletterSubmission } from 'libs/newsletter/src/models/newsletter.model';
-import { DaffNewsletterHubspotTransformer } from './newsletter.transformer';
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/transformers/newsletter.transformer.spec.ts]: Why this deletion?

File: libs/newsletter/src/driver/hubspot/hubspot.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import { HttpTestingController, HttpClientTestingModule } from '@angular/common/http/testing';
+import { TestBed } from '@angular/core/testing';
+import { DaffHubspotNewsletterService } from './newsletter.service';
+import { DaffNewsletterHubSpotDriverModule } from './hubspot-driver.module';
+import { of } from 'rxjs';
+
+describe('HubSpotDriver', () => {
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/hubspot.service.spec.ts]: This needs a better namespaced test descriptor.

File: libs/newsletter/src/driver/hubspot/models/config.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { Injectable } from '@angular/core';
+
+@Injectable({
+  providedIn: 'root'
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/models/config.ts]: Why is this injectable?

File: libs/newsletter/src/driver/hubspot/transformers/newsletter.transformer.spec.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+  });
+  it('should create the correct field values for a submission object', () => {
+
+    const payload: DaffNewsletterSubmission = { email: 'johndoe@email.com' };
+    const fieldValues = [{ 'name': 'email', 'value': 'johndoe@email.com' }];
+    expect(jsonBuilder(payload)).toEqual(fieldValues);
+  });
+  it('should create a proper HubspotRequest object', () => {
+
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/transformers/newsletter.transformer.spec.ts]: This spec doesn't look like it tests the code in question.

File: libs/newsletter/package.json (hunk 1 above)
Comment (Nolan-Arnold) [reply libs/newsletter/package.json]: It's not this was added trying to fix an import issue but I didn't remove it.

File: libs/newsletter/src/driver/hubspot/models/config.ts (hunk 4 above)
Comment (Nolan-Arnold) [reply libs/newsletter/src/driver/hubspot/models/config.ts]: To my understanding, it is Injectable for when it is called using `useValue` for the `DaffNewsletterConfig`

File: libs/newsletter/src/driver/hubspot/transformers/newsletter.transformer.spec.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { DaffNewsletterSubmission } from 'libs/newsletter/src/models/newsletter.model';
+import { DaffNewsletterHubspotTransformer } from './newsletter.transformer';
+import { jsonBuilder } from './json-builder';
+import { TestBed } from '@angular/core/testing';
+
+describe('DaffNewsletterHubspotTransformer', () => {
+  let transformer;
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/transformers/newsletter.transformer.spec.ts]: This needs to cover the `transformIn` function.

File: libs/newsletter/src/driver/hubspot/transformers/json-builder.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import { FieldValue } from '../models/hubspot-response'
+
+
+export const jsonBuilder = (payload: object): FieldValue[] => {
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/transformers/json-builder.ts]: Needs unit test.

File: libs/newsletter/src/driver/hubspot/hubspot.service.spec.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
+import { HttpTestingController, HttpClientTestingModule } from '@angular/common/http/testing';
+import { TestBed } from '@angular/core/testing';
+import { DaffHubspotNewsletterService } from './newsletter.service';
+import { DaffNewsletterHubSpotDriverModule } from './hubspot-driver.module';
+import { of } from 'rxjs';
+
+describe('DaffHubspotNewsletterDriver', () => {
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/hubspot.service.spec.ts]: This looks like a great test, but its an integration test, not a unit test. It should be moved to an integration test folder and you should specifically unit test the service in question.

Comment (Nolan-Arnold) [reply libs/newsletter/src/driver/hubspot/hubspot.service.spec.ts]: 78d941c09c0116264a0a369fe8e367926d2f78c8

**Comment by Nolan-Arnold** [comment]
I removed the spec `should send a request to a known Hubspot endpoint with a known configuration` as it overlapped with `should take Hubspot configuration from the module configuration`

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts [hunk 9]
```diff
@@ -59,16 +69,71 @@ describe('DaffHubspotNewsletterService', () => {
...
       req.flush('');
     })));
-  });
+    it('should include a the hubspot UTK in the request, if it hubspot cookie exists', () => {
+      doc.cookie = 'hubspotutk=mycookie;'
+      const submission = { email: 'email@email.com', firstName: 'John', lastName: 'Doe' };
+      newsletterService.send(submission).subscribe();
+
+      const req = httpMock.expectOne((req: HttpRequest<any>) => {
+        expect(req.url).toEqual('https://api.hsforms.com/submissions/v3/integration/submit/undefined/undefined');
```
Comment (Nolan-Arnold) [inline libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: These expects could be pushed out of the expectOne, let me know if this looks appropriate.

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+  afterEach(() => {
+    httpMock.verify();
+    document.cookie = 'hubspotutk=; expires=Thu, 01 Jan 1970 00:00:00 GMT'
+  });
+  it('should be created', () => {
+    expect(newsletterService).toBeTruthy();
+  });
+
+  describe('sending a newsletter submission to hubspot', () => {
+    it('should take Hubspot configuration from the module configuration', async(inject([HttpClient, DaffNewsletterConfig], (http, config) => {
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: This shouldn't use `async` and `inject`. It should be injected like the other services via the TestBed.

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+      spyOn(http, 'post').and.callThrough();
+      const submission = { email: 'email@email.com', firstName: 'John', lastName: 'Doe' };
+      const version = config.version === undefined ? 'v3' : config.version;
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: What does this code do?

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        + config.portalId + '/' + config.guid
+      newsletterService.send(submission).subscribe(() => {
+        expect(http.post).toHaveBeenCalledWith(submissionString, jasmine.any(Object));
+      });
+
+      const req = httpMock.expectOne('https://api.hsforms.com/submissions/v3/integration/submit/undefined/undefined');
+
+      req.flush('');
+    })));
+    it('should transform a request send to hubspot', async(inject([DaffNewsletterTransformer], (transformer) => {
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: sent*

Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: This shouldn't use async and inject. It should be injected like the other services via the TestBed.

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+      spyOn(transformer, 'transformOut').and.callThrough();
+      const submission = { email: 'email@email.com', firstName: 'John', lastName: 'Doe' };
+      newsletterService.send(submission).subscribe(() => {
+        expect(transformer.transformOut).toHaveBeenCalled();
+      });
+      const req = httpMock.expectOne('https://api.hsforms.com/submissions/v3/integration/submit/undefined/undefined');
+
+      req.flush('');
+    })));
+    it('should transform the response recieved from Hubspot', async(inject([DaffNewsletterTransformer], (transformer) => {
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: This shouldn't use async and inject. It should be injected like the other services via the TestBed.

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+      spyOn(transformer, 'transformIn').and.callThrough();
+      const submission = { email: 'email@email.com', firstName: 'John', lastName: 'Doe' };
+      newsletterService.send(submission).subscribe(() => {
+        expect(transformer.transformIn).toHaveBeenCalled();
+      });
+      const req = httpMock.expectOne('https://api.hsforms.com/submissions/v3/integration/submit/undefined/undefined');
+
+      req.flush('');
+    })));
+    it('should include a the hubspot UTK in the request, if it hubspot cookie exists', () => {
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: `should include the hubspot UTK in the request, if the hubspot cookie exists`

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+
+      const req = httpMock.expectOne((testReq: HttpRequest<any>) => {
+        expect(testReq.url).toEqual('https://api.hsforms.com/submissions/v3/integration/submit/undefined/undefined');
+        expect(testReq.body.context.hutk).toEqual('mycookie');
+        return testReq.url === 'https://api.hsforms.com/submissions/v3/integration/submit/undefined/undefined' && testReq.body.context.hutk === 'mycookie';
+      });
+
+      req.flush('');
+    });
+    it('should not include a the hubspot UTK in the request, if the hubspot cookie exists', () => {
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: `should not include the hubspot UTK in the request, if the hubspot cookie exists`

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+      newsletterService.send(submission).subscribe();
+
+      const req = httpMock.expectOne((testReq: HttpRequest<any>) => {
+        expect(testReq.url).toEqual('https://api.hsforms.com/submissions/v3/integration/submit/undefined/undefined');
+        expect(testReq.body.context.hutk).toEqual(null);
+        return testReq.url === 'https://api.hsforms.com/submissions/v3/integration/submit/undefined/undefined' && testReq.body.context.hutk === null;
+      });
+      req.flush('');
+    });
+    it('should include the pageUri in the request', async(inject([HttpClient, Router], (http, route) => {
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: This shouldn't use async and inject. It should be injected like the other services via the TestBed.

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+        expect(http.post).toHaveBeenCalledWith('https://api.hsforms.com/submissions/v3/integration/submit/undefined/undefined', jasmine.objectContaining({
+          'context': containingObject
+        }));
+      });
+
+      const req = httpMock.expectOne('https://api.hsforms.com/submissions/v3/integration/submit/undefined/undefined');
+
+      req.flush('');
+    })));
+    it('should include the pageName in the request', async(inject([HttpClient, Title], (http, title) => {
```
Comment (damienwebdev) [inline libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: This shouldn't use async and inject. It should be injected like the other services via the TestBed.

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts (hunk 11 above)
Comment (Nolan-Arnold) [reply libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: This code generates whats the URL is suppose to look like. So line 51 is logic to make sure a version is included in the mock submission URL.

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts (hunk 17 above)
Comment (Nolan-Arnold) [reply libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: Addressed in 7aeb11b1ed22f3bb2f03cf9a1023156c2824ca5b

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts (hunk 16 above)
Comment (Nolan-Arnold) [reply libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: Addressed in 7aeb11b1ed22f3bb2f03cf9a1023156c2824ca5b

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts (hunk 15 above)
Comment (Nolan-Arnold) [reply libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: Addressed in 7aeb11b1ed22f3bb2f03cf9a1023156c2824ca5b

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts (hunk 14 above)
Comment (Nolan-Arnold) [reply libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: Addressed in 7aeb11b1ed22f3bb2f03cf9a1023156c2824ca5b

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts (hunk 13 above)
Comment (Nolan-Arnold) [reply libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: Addressed in 7aeb11b1ed22f3bb2f03cf9a1023156c2824ca5b

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts (hunk 12 above)
Comment (Nolan-Arnold) [reply libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: Addressed in 7aeb11b1ed22f3bb2f03cf9a1023156c2824ca5b

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts (hunk 10 above)
Comment (Nolan-Arnold) [reply libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: Addressed in 7aeb11b1ed22f3bb2f03cf9a1023156c2824ca5b

File: libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts (hunk 11 above)
Comment (damienwebdev) [reply libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: Can we instead provide an explicit configuration object for `DaffNewsletterConfig ` instead of having some unnecessary testing logic?

Comment (Nolan-Arnold) [reply libs/newsletter/src/driver/hubspot/newsletter.service.spec.ts]: Addressed in 94d09a46797258af1bb71074d88eb9d7ca62fc86
